go/table.go

```diff
@@ -2,17 +2,37 @@ package flatbuffers
 
 // Table wraps a byte slice and provides read access to its data.
 //
-// The variable `Pos` indicates the root of the FlatBuffers object therein.
+// The byte slice `_buf` is not quite as it appears: We need to track a position
+// within the buffer (the root of the object that this Table represents), in
+// addition to the bounds of the buffer itself. Instead of using another field,
+// we put the position in the len of the slice, and the actual length in the
+// cap. When re-slicing it, `_buf` can be used as-is, but to avoid confusing
+// clients, we provide the Bytes() method to get the whole buffer.
 type Table struct {
-	Bytes []byte
-	Pos   UOffsetT // Always < 1<<31.
+	_buf []byte
+}
+
+// Init sets up the Table to refer to the object at offset `pos` in the buffer `buf`.
+// `pos` must be < 1<<31.
+func (t *Table) Init(buf []byte, pos UOffsetT) {
+	t._buf = buf[:pos:len(buf)]
```

Comment:
Can you explain this "double-colon" syntax to me please?

Reply:
Sure, it's called a "full slice expression" in the spec. (https://golang.org/ref/spec#Slice_expressions). The third argument becomes the cap of the new slice. Table never modifies the slice it's given or reads beyond its len, so there's no need to retain the original cap.

Reply:
Thanks! I was not familiar with that syntax. Bigger picture... what is the purpose of changing the capacity? It seems like this would be surprising to users, and make thinking about allocations trickier.

Reply:
I think I'm not explaining the idea clearly. Let me try again:

A Table needs to store three things, a buffer start, a buffer length, and a position within that buffer. The obvious way to do that is with a `[]byte` and an extra `int` field for the position. Because of alignment, that takes four words. But we only need three, can we drop one? Note that a `[]byte` has three words itself: a start, a length, and a capacity. We're only using two of them, the start and the length. The capacity is not used at all by Table. So what if we put the position there? Well, that doesn't work, because cap has to be greater than len. But we can swap them and put the position in the length field, and the length in the cap field. That means that the byte slice doesn't have its usual meaning, so we should keep clients from accessing it to avoid confusion. But it has all the information we need to make use of the Table, as long as we use cap as the end of each slice, instead of the default (len).

So, users don't have to think about it at all. They continue passing whatever `[]byte` they like to `GetRootAsX` or any `Init` function. The trick is encapsulated in Table, and Table never accesses anything beyond the length of the given slice.

Reply:
Interesting! So you're essentially using the slice header (in Go-speak) to store the metadata needed to access the data?

Reply:
Right, that's a good way to put it! It's using the fields of the slice header for something other than their intended meanings. So in a sense, the surprising thing is that it works without using `unsafe`. I think that's because the meanings are "close enough" to the intended meanings.